Sources/AsyncHTTPClient/ConnectionsState.swift

```diff
@@ -287,20 +261,12 @@ extension HTTP1ConnectionProvider {
 
         mutating func processNextWaiter() -> Action<ConnectionType> {
             if let waiter = self.waiters.popFirst() {
-                let (eventLoop, required) = self.resolvePreference(waiter.preference)
+                // There should be no case where we have waiters and available connections. As soon as connection
+                // is released and can be available, it should immideately go to the next waiter.
```

Comment:
Is this true in the face of event loop preferences? Is it not possible to have a waiter that is waiting for a connection on the appropriate event loop?

Reply:
Yes. We can only have waiters if there are no available connections, otherwise any new waiter will be leased a connection or a new connection will be created. If connection is released and it's active (meaning it could become "available"), it will be handled by a first branch in `release`. If the connection is not active, we will get here. But at this point the available invariant holds, released connection cannot become available, so the only course of action is to create a new one. Does that make sense?

Reply:
There are three other call sites for `processNextWaiter`:
1. `connectFailed`. This is somewhat similar - that failed connection will not be returned to the available set, since its already closed, so again, the only course of action is to create a new connection.
2. `remoteClose` - same thing
3. `timeout` - same thing as well

Reply:
We need to be extremely clear about our reasoning in this code. The comment reads: "As soon as connection is released and can be available, it should immideately go to the next waiter." I have natural questions about this: if the waiter has expressed a loop preference, it may be possible that a released connection is unsuitable for that waiter, despite being valid and active. What happens in this case?

It's totally possible that the _code_ is right, but the comment is at best incomplete.

Reply:
if the released connection is not suitable for the next waiter, it will be `.replace`d. How about the following comment instead:
```
This method is called in 4 cases:
1. from `release` when connection is inactive and cannot be re-used
2. from `connectFailed` when we failed to establish a new connection
3. from `remoteClose` when connection was closed by the remote side and cannot be re-used
4. from `timeout` when connection was closed due to idle timeout and cannot be re-used.
In all cases connection will not be `available`. Given that the waiter can only be present if there are no available connections (otherwise it would be leased a connection), where can be no situation where we can lease some available connection and the only course of action is to create a new one.
```

Reply:
> where can be no situation where we can

Reply:
```
This method is called in following cases:

1. from `release` when connection is inactive and cannot be re-used
2. from `connectFailed` when we failed to establish a new connection
3. from `remoteClose` when connection was closed by the remote side and cannot be re-used
4. from `timeout` when connection was closed due to idle timeout and cannot be re-used.

In all cases connection, which triggered the transition, will not be in `available` state.

Given that the waiter can only be present in the pool if there were no available connections 
(otherwise it would be leased a connection immediately), we do not see a situation when we can lease another
available connection, therefore the only course of action is to create a new connection for the waiter.
```

Complex comment highlights that logic is convoluted here, but I am not ready to propose an immediate improvement for the logic.

Reply:
Happy for the complex comment to go in.